File: ruby/trema/match.c
/*
 * Copyright (C) 2008-2013 NEC Corporation
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2, as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */


#include "trema.h"
#include "ruby.h"
#include "conversion-util.h"


extern VALUE mTrema;
VALUE cMatch;


/*
 * Creates a {Match} instance from packet_in's data.
 *
 * @overload match_from(message)
 *
 *   @example
 *     def packet_in datapath_id, message
 *       send_flow_mod(
 *         datapath_id,
 *         :match => Match.from( message ),
 *         :actions => Trema::ActionOutput.new( 2 )
 *       )
 *     end
 *
 *   @param [PacketIn] message
 *     the {PacketIn}'s message content.
 *
 * @return [Match] self
 *   the match from packet_in's data.
 */
static VALUE
match_from( VALUE self, VALUE message ) {
  UNUSED( self );
  uint32_t in_port = OFPP_CONTROLLER;
  VALUE r_match = rb_iv_get( message, "@match" );
  if ( !NIL_P( r_match ) ) {
    VALUE r_in_port = rb_iv_get( r_match, "@in_port" );
    if ( !NIL_P( r_in_port ) ) {
      in_port = NUM2UINT( r_in_port );
      if ( in_port == 0 ) {
        rb_raise( rb_eArgError, "The in_port value must be greater than 0." );
      }
      if ( in_port > OFPP_MAX ) {
        if ( in_port != OFPP_CONTROLLER && in_port != OFPP_LOCAL ) {
          rb_raise( rb_eArgError, "The in_port value must be less than or equal to OFPP_MAX." );
        }
      }
    }
  }

  VALUE r_data = rb_iv_get( message, "@data" );
  if ( NIL_P( r_data ) ) {
    rb_raise( rb_eArgError, "The data is a mandatory option" );
  }
  buffer *data = r_array_to_buffer( r_data );
  bool ret = parse_packet( data );
  if ( !ret ) {
    rb_raise( rb_eArgError, "The data must be a Ethernet frame." );
  }

  oxm_matches *matches = create_oxm_matches();
  set_match_from_packet( matches, in_port, NULL, data );
  VALUE r_new = oxm_match_to_r_match( matches );
  delete_oxm_matches( matches );

  free_buffer( data );

  return r_new;
}


void
Init_match( void ) {
  cMatch = rb_define_class_under( mTrema, "Match", rb_eval_string( "Trema::Message" ) );
  rb_require( "trema/match" );
  rb_define_singleton_method( cMatch, "from", match_from, 1 );
}


/*
 * Local variables:
 * c-basic-offset: 2
 * indent-tabs-mode: nil
 * End:
 */

File: ruby/trema/match.rb
#
# Copyright (C) 2008-2013 NEC Corporation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License, version 2, as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along
# with this program; if not, write to the Free Software Foundation, Inc.,
# 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
#

module Trema
  class Match < Message
    unsigned_int32 :in_port, :in_phy_port
    unsigned_int64 :metadata, :metadata_mask
    mac :eth_src, :eth_src_mask
    mac :eth_dst, :eth_dst_mask
    unsigned_int16 :eth_type
    unsigned_int16 :vlan_vid, :vlan_vid_mask
    unsigned_int8 :vlan_pcp
    unsigned_int8 :ip_dscp # IP DSCP ( 6 bits in ToS field )
    unsigned_int8 :ip_ecn # IP ECN ( 2 bits in ToS field )
    unsigned_int8 :ip_proto # ip protocol
    ip_addr :ipv4_src, :ipv4_src_mask
    ip_addr :ipv4_dst, :ipv4_dst_mask
    unsigned_int16 :tcp_src, :tcp_dst
    unsigned_int16 :udp_src, :udp_dst
    unsigned_int16 :sctp_src, :sctp_dst
    unsigned_int8 :icmpv4_type, :icmpv4_code
    unsigned_int16 :arp_op
    ip_addr :arp_spa, :arp_spa_mask
    ip_addr :arp_tpa, :arp_tpa_mask
    mac :arp_sha, :arp_sha_mask
    mac :arp_tha, :arp_tha_mask
    ip_addr :ipv6_src, :ipv6_src_mask
    ip_addr :ipv6_dst, :ipv6_dst_mask
    unsigned_int32 :ipv6_flabel, :ipv6_flabel_mask
    unsigned_int8 :icmpv6_type, :icmpv6_code
    ip_addr :ipv6_nd_target
    mac :ipv6_nd_sll, :ipv6_nd_tll
    unsigned_int32 :mpls_label
    unsigned_int8 :mpls_tc, :mpls_bos
    unsigned_int32 :pbb_isid, :pbb_isid_mask
    unsigned_int64 :tunnel_id, :tunnel_id_mask
    unsigned_int16 :ipv6_exthdr, :ipv6_exthdr_mask
  end
end

### Local variables:
### mode: Ruby
### coding: utf-8-unix
### indent-tabs-mode: nil
### End:
